lib/backend/test/atomicwrite.go

```diff
@@ -0,0 +1,688 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+// AtomicWriteConstructor is equivalent to [Constructor], except that it includes the new AtomicWrite method. This type
+// will be deprecated once all backends implement AtomicWrite.
+type AtomicWriteConstructor func(options ...ConstructionOption) (backend.AtomicWriteBackend, clockwork.FakeClock, error)
+
+func RunAtomicWriteComplianceSuite(t *testing.T, newBackend AtomicWriteConstructor) {
+	t.Run("Move", func(t *testing.T) {
+		testAtomicWriteMove(t, newBackend)
+	})
+
+	t.Run("Lock", func(t *testing.T) {
+		testAtomicWriteLock(t, newBackend)
+	})
+
+	t.Run("Max", func(t *testing.T) {
+		testAtomicWriteMax(t, newBackend)
+	})
+
+	t.Run("Concurrent", func(t *testing.T) {
+		testAtomicWriteConcurrent(t, newBackend)
+	})
+
+	t.Run("NonConflicting", func(t *testing.T) {
+		testAtomicWriteNonConflicting(t, newBackend)
+	})
+
+	t.Run("Other", func(t *testing.T) {
+		testAtomicWriteOther(t, newBackend)
+	})
+}
+
+// testAtomicWriteMove verifies the correct behavior of "move" operations.
+func testAtomicWriteMove(t *testing.T, newBackend AtomicWriteConstructor) {
+	bk, _, err := newBackend()
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	fromKey, toKey, val := []byte("/src"), []byte("/dest"), []byte("val")
+
+	lease, err := bk.Put(ctx, backend.Item{
+		Key:   fromKey,
+		Value: val,
+	})
+	require.NoError(t, err)
+
+	// perform "move".
+	_, err = bk.AtomicWrite(ctx, []backend.ConditionalAction{
+		{
+			Key:       fromKey,
+			Condition: backend.Revision(lease.Revision),
+			Action:    backend.Delete(),
+		},
+		{
+			Key:       toKey,
+			Condition: backend.Whatever(),
+			Action: backend.Put(backend.Item{
+				Value: val,
+			}),
+		},
+	}...)
```

Comment:
Any particular reason?  I find it easier to quickly do a visual assessment of what an `AtomicWrite` call is doing when using the slice syntax since there is less visual noise inside the main body of the args.

Reply:
+1, making it a slice argument makes it even more explicit that it's a multi-item operation too.